### ISSUE TITLE
Move all repositories one Monolithic repo?

### ISSUE THREAD

**Issue by jaapio** [phpDocumentor/phpDocumentor2#1945]
In this issue I would like to discuss the proposal of @TomasVotruba to move all libraries to a Monolithic repository. To restructure our CI work. Tomas wrote a [blog post](https://www.tomasvotruba.cz/blog/2017/01/31/how-monolithic-repository-in-open-source-saved-my-laziness/#example-how-it-s-done-in-symplify) about this.

I'm not fully convinced that a monolithic repository is the way to go. However symfony seems to use it for years. On the other side the zend framework team stopped this way of working since the had a lot of needless releases because not all components are updated at the same time. 

I would like to investigate what the impact would be on the project to work in this style. And find out what the thoughts of the team are about this approach. 

I would like to hear your opinions about this.

**Comment by jaapio** [comment]
Related issues
- https://github.com/phpDocumentor/ReflectionDocBlock/pull/145
- https://github.com/phpDocumentor/fig-standards/issues/114

**Comment by mvriel** [comment]
Personally; not a fan of it. But honestly: I haven't worked in a setup like that so far so I may be wrong. Nonetheless, the problem with it is that it is hard to have separate maintainers for it and you need to keep the split repository up to date. 

It sounds easy to forget that and while your project has all the benefits, your dependents do not.

Even the mentioned advantage of bumping all components to a new version of PHP; i.m.o. this should not be done automatically but should be a conscious decision per component.

Also: how does this work with the `composer.json`? Would each component still have a separate `composer.json`?

As you can probably infer from the above; I am quite skeptical. But I would love to hear more on the subject. The blog post was not sufficient enough for me to get a good grasp of the trade-offs.

**Comment by jaapio** [comment]
I share your skeptics, I see that zend framework has members that are owner of a specific component now the split up. They allowed the community maintain some of the components where symfony has a small core team that is allowed to push to the repo. 

We are not a community of the size like symfony and zend so that might not apply fully. 

I know that zend had the composer.json files in the main repo. But it doesn't allow you to have lock files of for all of them. I never understood how they managed these composer.json files.

On the other hand I never understood which repo I had to create issues on. People using the splitted component always had to search for the right way to create their issues. 

I don't belief that development on doctrine or other libraries slows down because of the complexity of their maintenance. I think it is more a matter of time. The opposite was proven by zend framework where the redesign of components was blocked by the tight connection between components. Some components are not even usable without others. I think we are doing this better. since the type resolver is really independent from others for example. 

The only advantage I see is that we have to maintain some configuration once, like travis and phpstan. The code quality will be at the same level over all components. However I would expect that you still need to run unittest on the components. Because you want to make sure components are correctly working? 

I don't have any experience with the monolithic approach. But I'm also not convinced by the blog post only. And I think it might work very well when you are working alone. But the benefits are gone when your community grows. 

Only thing I have in mind right now it that we could be more selective at what we are pushing to a separate repo. I think flyfinder and our reflection components are good examples of packages that could have been part of the main repository. Since the are developed for phpdocumentor. And I think it is very unlikely that others will start using them. On the other hand having them in the main repo wouldn't even allow separate usage. For that purpose Monolithic approach could be a solution 

I don't want to close the door to the monolithic repo right now. And I think it is good to have a discussion about the topic.

**Comment by mvriel** [comment]
Small remark: the reflection components are used by some dependents; PHPUnit, Wordpress and (IIRC) Better Reflection come to mind

**Comment by jaapio** [comment]
They do use reflection-docblock and dependencies, but not reflection, since they want to parse docblocks / annotations. To create behavior like our reflection component which is quite limited from a reflection point of view.

**Comment by TomasVotruba** [comment]
I had similar concerns untill I tried this and postponed this technique for years just because of lack of knowledge and fear of unknown.

I could write many arguments like I did before, but the most working was real example.

Would you be open to monorepo demo for PhpDocumentor?

If so, what repositories should be there?

**Comment by cdp1337** [comment]
I authored a framework that we use internally and took the same approach; a monolithic repo.  In my example, that ended up *not* being the most optimal of setups.  Since it is a framework akin to ZF or Symphony, it has a central core and multiple components that can work together that may or may not be included in any given project.  If I had a client site that required the Core and components A, B, and C, I needed scripts to manually extract the required components and merge them in, this means I lost any/all benefits from git changelogs, as they were truly external code repos at that stage.

Additionally, if everything is in one monolithic repo, you can't divvy permissions.  For a fully FOSS project, this may not be a huge deal, but you still may not want or need an individual working on component A to push changes for component B, so keeping them separated helps logistically in that case.  (Plus the added benefit that you could do protected parts, should you need to.)

Last noteworthy point that was the tipping point for my project; issue tracking.  We ran into logical pains where if a problem with component A was found, it was recorded on the parent group, which usually did not have any reason to contain a log of the nested issue.  Additionally, if another project used the same component, it would often have duplicated issues because users would run into the same problem across different projects.  Even worse, if two developers fixed the same problem across two different projects, then you'd have to fight with code overwrites, because you don't even know if there's a conflict!

Using individual repos also gave me the ability to *still* have a "monolithic project" should i want to; simply by importing all the required projects as git submodules!  This provides developers the ability to work across multiple projects _almost_ seamlessly.

---

TLDR; Monolithic < individualized repos and git:submodules, in my use case.

**Comment by TomasVotruba** [comment]
@cdp1337 Thanks for sharing your experience. How do you moved to split repos?

**Comment by cdp1337** [comment]
Painfully truth be told.  I had to create individual repos for each component, then manually merge in the files from that corresponding component into the new repo.  Once I got some of the oddities of submodules nailed down though, I was able to streamline the process of repo creation and submdodule embedding with a nice script.  In my case it scans git repos for any compatible repo and prompts the developer which ones they want to "install" into their project, then does the necessary git submodule add commands to grab the code.  It even has an uninstall option where developers can remove submodules from a linked project.

https://git.eval.bz/coreplus/core/blob/master/setup.sh#L437

for an example of my "setup" script that handles the submodule linking and unlinking, if you're curious.  I haven't gotten a chance to merge up to github yet, it's on my todo list.

**Comment by TomasVotruba** [comment]
I see. Do you have feedback from other devs in the team? 
How many people work with that?

**Comment by mvriel** [comment]
To keep momentum with the development of phpDocumentor and because I do not believe that we encounter any issues (at least not those that hurt enough or would be solved with a monolithic repo) from our current way of working; I am going to close this issue.

Closing in this instance does not mean that it is a bad suggestion or that we completely disregard this option. It is in this instance a way of saying 'at least not now' and closing this item will help in managing our list of open issues.

The discussion is good, and perhaps we can later restart it when there is more leeway.

**Comment by TomasVotruba** [comment]
Thanks for nice summary Mike, I agree :+1: 

For future reference: there is now a tool that makes setup + split in 3 yaml lines:
https://github.com/symplify/monorepobuilder#5-split-directories-to-git-repositories

Feel free to use it :+1:
